app/main.py

```diff
@@ -1,2 +1,204 @@
+
+    def __init__(self, capacity: int = INITIAL_CAPACITY) -> None:
+        self.capacity = capacity
+        self.size = 0
+        self.hash_table: list[Dictionary.Node | None] = [None] * self.capacity
+
+    def _calculate_index(self, key: Hashable) -> int:
+        index = hash(key) % self.capacity
+
+        while (
+            self.hash_table[index] is not None
+            and self.hash_table[index].key != key
+        ):
+            index = (index + 1) % self.capacity
+
+        return index
+
+    @property
+    def current_max_size(self) -> int:
+        return int(self.capacity * self.RESIZE_THRESHOLD)
+
+    def _resize(self) -> None:
+        old_hash_table = self.hash_table
+
+        self.__init__(self.capacity * self.CAPACITY_MULTIPLIER)
+
+        for node in old_hash_table:
+            if node is not None:
+                self.__setitem__(node.key, node.value)
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        index = self._calculate_index(key)
+        key_hash = hash(key)
+
+        if self.hash_table[index] is None:
+            if self.size + 1 >= self.current_max_size:
+                self._resize()
+                index = self._calculate_index(key)
+            self.size += 1
+
+        self.hash_table[index] = Dictionary.Node(key, value, key_hash)
+
+    def __getitem__(self, key: Hashable) -> Any:
+        index = self._calculate_index(key)
+
+        if self.hash_table[index] is None:
+            raise KeyError(f"Cannot find value for key: {key}")
+
+        return self.hash_table[index].value
+
+    def __delitem__(self, key: Hashable) -> None:
+        index = self._calculate_index(key)
+
+        if self.hash_table[index] is None:
+            raise KeyError(f"Cannot delete value for key: {key}")
+
+        self.hash_table[index] = None
+        self.size -= 1
+
+    def __iter__(self) -> Iterator[tuple[Hashable, Any]]:
+        for node in self.hash_table:
+            if node is not None:
+                yield node.key, node.value
+
+    def __len__(self) -> int:
+        return self.size
+
+    def __str__(self) -> str:
+        items = [
+            f"{node.key}: {node.value}"
+            for node in self.hash_table
+            if node is not None
+        ]
+        return "{" + ", ".join(items) + "}"
+
+    def get(self, key: Hashable, default: Optional[Any] = None) -> Any:
+        """
+        Get the value for the given key.
+        If the key does not exist, return the default value.
+
+        Default value is optional.
+        If it was not provided, method will raise KeyError.
+        """
+        try:
+            return self[key]
+        except KeyError:
+            if default is not None:
+                return default
+            else:
+                raise KeyError(
+                    f"Cannot find value for key: {key}"
+                    "and no default value is provided"
+                )
+
+    def pop(self, key: Hashable) -> Any:
```

Comment:
pop should also have a default param but it behaves a bit different from `get` when default value is not set and key is not found:
Dictionary.pop(key, None) should return None instead of triggering KeyError
Dictionary.pop(key) should raise KeyError